Revamp integration,streaming quick start guides (#4592)
wso2/docs-apim: en/docs/get-started/streaming-quick-start-guide.md

# Quick Start Guide - Streaming Integration

Let's get started with WSO2 Streaming Integrator(SI) by running a simple streaming use case in your local environment. 

### Before you begin...

1. Install [Oracle Java SE Development Kit (JDK)](http://java.sun.com/javase/downloads/index.jsp) version 11 or 1.8 and set the `JAVA_HOME` environment variable.
   For more information on setting the `JAVA_HOME` environment variable for different operating systems, see [Setup and Install]({{base_path}}/install-and-setup/install/installing-the-product/installing-api-m-runtime/).
2. Download the Streaming Integrator and Streaming Integrator Tooling distributions from the [WSO2 Streaming Integrator site](https://wso2.com/integration/streaming-integrator/) and extract them to a location of your choice. Hereafter, the extracted location is referred to as `<SI_HOME>` and `<SIT_HOME>` respectively.<br/><br/>

### What you'll build

> In this sample scenario, you will aggregate the data relating to the raw material purchases of a sweet production factory and publishers the data to a WebSocket receiver.

Reviewer: ```suggestion
In this sample scenario, you aggregate the data relating to the raw material purchases of a sweet production factory and publish the data to a WebSocket receiver.
```

Author: fixed.

                                

### Step 1: Start the Streaming Integrator

To start WSO2 Streaming Integrator, navigate to the `<SI_HOME>/bin` directory from the CLI, and issue the appropriate command based on your operating system:

- **For Linux**: `./server.sh`
- **For Windows**: `server.bat --run`

### Step 2: Start the Streaming Integrator Tooling
To start WSO2 Streaming Integrator Tooling, navigate to the `<SIT_HOME>/samples/sample-clients/websocket-receiver` directory from the CLI, and issue `ant`.

> This will 

Reviewer: This sentence seems missing.

Author: fixed

- **For Linux**: `./server.sh`

> - **For Windows**: `server.bat --run`

Reviewer 2: There seems to be something missing here. Coz it says issue `ant` and then also gives the correct server startup commands.

Author: fixed

### Step 3: Create and deploy a simple Siddhi application

Let's create a simple Siddhi application that reads data from a XML file, does a simple transformation to the data, and then writes the results to the WebSocket receiver running.

1. Download `sampledata.xml` file from [here]({{base_path}}/assets/attachments/quick-start-guide/sampledata.xml) and save it in a location of your choice.

    !!! info
        In this example, the file is located in the `/Users/foo` directory.

2. Open a text file and copy-paste the following Siddhi application into it.

    !!! tip

>         Here, a sample Siddhi application is provided to minimize the time spend following this guide. But we recommend using Streaming Integration Tooling that offers features such as syntax checking, event simulation for testing purposes, reformatting code, the option to design applications in a graphical interface or by writing code, and many more. For more information to design the Siddhi applications, see [Streaming Integrator Tooling Overview]({{base_path}}/develop/streaming-apps/streaming-integrator-studio-overview).

Reviewer 2: ```suggestion
        Here, a sample Siddhi application is provided to minimize the time spend following this guide. However, WSO2 recommends that you use the Streaming Integration Tooling that offers features such as syntax checking, event simulation for testing purposes, reformatting code, the option to design applications in a graphical interface or by writing code, and many more. For more information on designing Siddhi applications, see [Streaming Integrator Tooling Overview]({{base_path}}/develop/streaming-apps/streaming-integrator-studio-overview).
```

Author: fixed

Reviewer: Change "spend" to "spent".

Author: Fixed

    ```
    @App:name('ManageStats')
    @App:description('Receive events via a xml file and write sum of the amount to websocket')

    @source(type = 'file', mode = "TEXT.FULL", file.uri = "file:/Users/foo/sampledata.xml",
    @map(type = 'xml'), enclosing.element='/events/event',enable.streaming.xml.content='true',
    @attributes(name = "name", amount = "amount"))
    define stream TestProductionStream (name string, amount double);

    @sink(type='websocket', url='ws://localhost:8025/abc',
    @map(type='xml'))
    @sink(type='log')
    define stream ProductionSumAlertStream (name string, amount double);

    -- Simple Siddhi query to calculate production totals.
    @info(name = 'query1')
    from TestProductionStream
    select name, sum(amount) as the amount
    group by name
    insert into ProductionSumAlertStream;
    ```

3. Save this file as `ManageStats.siddhi` in the `<SI_HOME>/wso2/server/deployment/siddhi-files` directory.

    This deploys the `ManageStats` in the Streaming Integrator. The following message appears to indicate that the Siddhi application installed successfully.

>     `INFO {org.wso2.carbon.siddhi.editor.core.internal.WorkspaceDeployer} - Siddhi App ManageStats.siddhi successfully deployed.`

Reviewer 2: ```suggestion
    ```INFO {org.wso2.carbon.siddhi.editor.core.internal.WorkspaceDeployer} - Siddhi App ManageStats.siddhi successfully deployed.```
```

Author: fixed

## Step 4: Test your Siddhi application

To test the `ManageStats` Siddhi application you created, follow the steps below.

1. Open the `/Users/foo/sampledata.xml` file and add rows in it as follows:

    `    
    <event>
        <name>Almond cookie</name>
        <amount>70</amount>
    </event>
    <event>
        <name>Baked alaska</name>
        <amount>30</amount>
    </event>
    <event>
        <name>Toffee</name>
        <amount>60</amount>
    </event>`

    Save your changes.

> 2. To observe the MI console output.

Reviewer 2: ```suggestion
2. Observe the MI console output.
```

Author: fixed

> 2. To observe the MI console output.

Reviewer 2: Should this be "MI console output" OR "SI console output"?

Author: This shpuld be MI. the SiddhiApp is deployed in MI.

You can see the following message in the SI console log.

> ![]({{base_path}}/assets/img/streaming/qsg/gsg-streaming-output.png)

Reviewer 2: ```suggestion
[![QSG streaming output]({{base_path}}/assets/img/streaming/qsg/gsg-streaming-output.png)]({{base_path}}/assets/img/streaming/qsg/gsg-streaming-output.png)
```

Author: fixed.

Reviewer: Shall we make this text instead of an image. Text is better from a maintenance perspective.

Author: fixed

# Quick Start Guide - Streaming Integration

Let's get started with WSO2 Streaming Integrator(SI) by running a simple streaming use case in your local environment. 

## Before you begin...

1. Install [Oracle Java SE Development Kit (JDK)](http://java.sun.com/javase/downloads/index.jsp) version 11 or 1.8 and set the `JAVA_HOME` environment variable.
   For more information on setting the `JAVA_HOME` environment variable for different operating systems, see [Setup and Install]({{base_path}}/install-and-setup/install/installing-the-product/installing-api-m-runtime/).

> 2. Download the Streaming Integrator and Streaming Integrator Tooling distributions from the [WSO2 Streaming Integrator site](https://wso2.com/integration/streaming-integrator/) and extract them to a location of your choice. Hereafter, the extracted location is referred to as `<SI_HOME>` and `<SIT_HOME>` respectively.<br/><br/>

Reviewer: Why are we downloading SI and not API-M?

Author: yes

     @source(type = 'file', mode = "TEXT.FULL", dir.uri = 'file:/Users/foo/files', enclosing.element = "/events/event", enable.streaming.xml.content = "true", 
    	@attributes(name = "name", amount = "amount"),
	    @map(type = 'xml'))
     @sink(type = 'log')
    define stream SweetsProductionStream (name string, amount double);

    @sink(type = 'websocket-server', host = "localhost", port = "8025",
	    @map(type = 'json'))
    define stream ProductionAlertStream (name string, amount double);

    @sink(type = 'log')
    define stream LowProductionAlertStream (name string, amount double);

     -- Simple Siddhi query to calculate production totals.
     @info(name = 'query1')
    from SweetsProductionStream
    select name, sum(amount) as amount 
	    group by name 
    insert into ProductionAlertStream;
    ```
    
3. Save this file as `ManageProductionStats.siddhi`.

4. Click on **Deploy** and then **Deploy To Server** to deploy the Siddhi Application in Streaming Integrator.

4. Add Streaming Integrator server details by clicking on `Add New Server`

    Specify the Streaming Integrator host `localhost` and port `9443`.
    

> 5. Select the `ManageProductionStats` and the Server and **Deploy**.

Reviewer: Is it Server URL or Server Name?

Author: it is a Server  we have added in the above step to select.

Reviewer: Ok, let's specify that.

The following message appears to indicate that the Siddhi application deployed successfully in the Streaming Integrator console.

    ```INFO {org.wso2.carbon.streaming.integrator.core.internal.StreamProcessorService} - Siddhi App ManageProductionStats1 deployed successfully```

### Step 4: Test your Siddhi application

> To test the `ManageProductionStats` Siddhi application above created, follow the steps below.

Reviewer: ```suggestion
To test the `ManageProductionStats` Siddhi application created above, follow the steps below.
```

Author: fixed

1. Copy the `sampledata.xml` file in the `/Users/foo/files` directory. 

2. Observe the SI console output.

You can see the following message in the SI console log.

```jvm
    INFO {io.siddhi.core.stream.output.sink.LogSink} - ManageProductionStats1 : ProductionAlertStream : [Event{timestamp=1630491334294, data=[Almond cookie, 170.0], isExpired=false}, Event{timestamp=1630491334294, data=[Baked alaska, 100.0], isExpired=false}, Event{timestamp=1630491334294, data=[Toffee, 100.0], isExpired=false}] 
```

</br>
 Congratulations!

>  Now, you have created a your first Streaming service. You may follow steps given below to expose the service as a Managed API in API Manager.

Reviewer: ```suggestion
 Now, you have created a your first Streaming service. Follow the steps given below to expose the service as a Managed API in API Manager.
```

Author: fixed

      
## Exposing an Streaming Service as a Managed API

The `ManageProductionStats` Siddhi Application you deployed in the Micro Integrator is a **streaming service** for the API Manager. Now, let's look at how you can expose the streaming service to the API Management layer and generate a managed API by using the service.

### Step 1 - Expose as a service 

1. Start the API Manager runtime:

    1.  Go to the [WSO2 API Manager website](https://wso2.com/api-management/), click **TRY IT NOW**, and then click **Zip Archive** to download the API Manager distribution as a ZIP file.

>     2.  Extract the ZIP file.

Reviewer: These should be added to the Before you Begin section for consistency. We seem to have the download instructions there.

Author: because we need tis only if these spes are following.
Still need to move there as optional downloads?

Reviewer: Yes, let's move it anyway. If SI is packaged inside it, it should be a mandatory download instead of SI.

Author: moved.

         ```
    
    3.  Start the Streaming Integrator again.

3. Access the streaming service from the **API Publisher**:

    1.  Sign in to the API Publisher portal: `https://localhost:9443/publisher`. 

    !!! Tip
        Use `admin` as the user name and password.

    2.  You can also click the **hamburger** icon on the upper-left and click **Services** to see the available services.

        <img src="{{base_path}}/assets/img/integrate/tutorials/service-catalog/open-service-catalog.png" width="400">
  
    3. See that the `SweetProductionApplication` is listed as a service.

### Step 2 - Create a managed API using the Streaming Service

1.  Click on the `SweetProductionApplication` service.

2.  Click **Create API**.

     This opens the **Create API** dialog box with the API details that are generated based on the service.

    <a href="{{base_path}}/assets/img/streaming/qsg/create-api-from-streaming-service.png"><img src="{{base_path}}/assets/img/streaming/qsg/create-api-from-streaming-service.png" width="800"></a>

3.  Update the API name, context, and version if required, and click **Create API**. 

    The overview page of the API that you just created appears. 

> 4. Optionally, Update the portal configurations and API configurations as required.

Reviewer: ```suggestion
4. Optionally, update the portal configurations and API configurations as required.
```

     Now, you have successfully created a Web Socket API using the service.

### Step 3 - Publish the managed API

> 1. Navigate  to **Deployments** and click **Deploy** to create a revision to deploy in the default Gateway environment. 

Reviewer: ```suggestion
1. Navigate to **Deployments** and click **Deploy** to create a revision to deploy in the default Gateway environment. 
```

> 2. Navigate to **Lifecycle** and click  **Publish** to publish the API in the Gateway environment.

Reviewer: ```suggestion
2. Navigate to **Lifecycle** and click **Publish** to publish the API in the Gateway environment.
```

    
    If the API is published successfully, the lifecycle state will shift to **PUBLISHED**. 

### Step 4 - Invoke the Managed API via Developer Portal

> 1. Navigate to  the **Developer Portal** by clicking on the `View In Dev Portal` at the top menu.

Reviewer: ```suggestion
1. Navigate to  the **Developer Portal** by clicking on `View In Dev Portal` at the top menu.
```

Author: fixed

2. Sign in using the default username/password `admin/admin`. You will be redirected to the **APIs**.

3. Under **APIs**, you will see the published `SweetProductionApplication`. Click on it to navigate to the Overview of the API.

> 4. Click `Subscriptions` on the left Menu and click `SUBSCRIBE` This will create a subscription to the API using the `DefaultApplication`.

Author: fixed

    <a href="{{base_path}}/assets/img/streaming/qsg/streaming-api-subscribe.png"><img src="{{base_path}}/assets/img/streaming/qsg/streaming-api-subscribe.png.png" size="800"></a>

5. Click `PROD KEYS` to generate keys for the application and click `GENERATE  KEYS`.
    
    <a href="{{base_path}}/assets/img/streaming/qsg/streaming-api-subscribe.png"><img src="{{base_path}}/assets/img/streaming/qsg/streaming-api-subscribe.png" size="800"></a>

> 6. Copy the generated access token for future use. Alternatively you can generate a new access token by using `GENERATE ACCESS TOKEN`.

Author: fixed.

> 7. Click `Try Out` on the left Menu Cllick on the **SUB** topic.

Author: fixed.
